### PR TITLE
RfC: Use `todo!()` instead of `()` for missing fields

### PR DESCRIPTION
**Pull request by jo-so**
Most commonly a field of a struct can be initialized with its default value than an empty tuple.

### CODE REVIEW

File: crates/ide_diagnostics/src/handlers/missing_fields.rs [hunk 1]
```diff
@@ -167,7 +167,7 @@ fn main() {
 pub struct Foo { pub a: i32, pub b: i32 }
 "#,
             r#"
-fn some(, b: () ) {}
+fn some(, b: Default::default() ) {}
```
Comment (bjorn3) [inline crates/ide_diagnostics/src/handlers/missing_fields.rs:170]: This is wrong.

**Comment by bjorn3** [comment]
I think it is harded to replace `Default::default()` than `()`. Could you at least check that the field type implements `Default` and use `vec![]` for `Vec` and `None` for `Option`? Maybe it would also be nice to use the only possible value directly for unit structs?

File: crates/ide_diagnostics/src/handlers/missing_fields.rs (hunk 1 above)
Comment (jo-so) [reply crates/ide_diagnostics/src/handlers/missing_fields.rs:170]: I think that's what the comment above says: `FIXME: this is very wrong, but somewhat tricky to fix.`

**Comment by jo-so** [comment]
But with `Default::default()` the code compiles most of the time while it doesn't with `()`. And leaving `Default::default()` makes it easier when you refactor the structure, e.g. you change a `Vec<…>` into a `HashSet<…>`.

Maybe someone else can tell how he thinks about `Default::default()` vs. `()`.

**Comment by Veykril** [comment]
We should check if it implements the necessary trait certainly, and ideally also make this a config option as replacing that default path expression is a lot more work than replacing `()` so I'm sure there are people who prefer `()` over that. cc https://github.com/rust-analyzer/rust-analyzer/issues/1470

File: crates/ide_diagnostics/src/handlers/missing_fields.rs (hunk 1 above)
Comment (Veykril) [reply crates/ide_diagnostics/src/handlers/missing_fields.rs:170]: Ye that's the macro at fault, nothing specific to this PR

**Comment by matklad** [comment]
This should be consitent with `fill_match_arms` first and foremost. That uses `todo!()` at the moment.

Either way, we should extract the code for "come up with default expression for type" into assist/utils, and use that in both places. 


As to what's that expression is to be, I see the following choices:

* `()` as the smallest thing
* `todo!()` short, and semantically correct thing
* default value for a type (`0`, `None`, `HashMap::default()`, depending on the type in question).

I think `todo!()` is probably the best  choice -- it's reasonably compact, reasonably correct, and robust (it doesnt' depend on type of the thing, which is an advantage, as IDE always behaves predictably).

**Comment by matklad** [comment]
relevant bit from fill match arms: https://github.com/rust-analyzer/rust-analyzer/blob/25368d24308d6a94ffe8b99f0122bcf5a2175322/crates/ide_assists/src/handlers/fill_match_arms.rs#L132

**Comment by Veykril** [comment]
@jo-so are you interested in trying to implement this configurably as outlined by matklad? If yes feel free to ask questions here or on zulip, if not that's also fine!

**Comment by matklad** [comment]
To clarify -- I don't think we need configurability in *this* case, we should just pick a good choice. I am confident that `todo!()` is a reasonable choice, but if folks feel strongly that some other choice would be better, than yeah, we can make this configurable!

**Comment by flodiebold** [comment]
I would prefer `todo!()` and kind of want to use `todo!()` as 'holes' with some editor functionality to switch between them and replace them easily.

**Comment by bjorn3** [comment]
XCode has native support for placeholders. They are automatically inserted for arguments when autocompleting. They are saved as `<#placeholder text#>` and appear as blue rounded rectangles that can only be selected as a whole. You can navigate between them using tab and shift+tab.

Maybe it is possible to emulate it using the javascript api of vscode? Or do lsp replacement inlay hints act as a single selectable entity in vscode? As for the file syntax, while `<#placeholder#>` may be valid inside macros, I think it is sufficiently uncommon that there is value in using this format just like XCode.

**Comment by ruabmbua** [comment]
First random thought on this: why not use todo!() instead? Will look as ugly as Default::default(), but work in (probably) all cases.

**Comment by jo-so** [comment]
I've found a way to do this in Emacs. So, I'm fine with everything even dropping the PR.

**Comment by lnicola** [comment]
changelog fix (first contribution) use `todo!()` instead of `()` in "Fill missing fields"

**Pull request merged**